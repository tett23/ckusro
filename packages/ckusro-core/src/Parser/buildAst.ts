import { Content, Parent, Root } from 'mdast';
import { FileBuffer, newDoesNotExistFile } from '../models/FileBuffer';
import { compareInternalPath } from '../models/InternalPath';
import { Plugins } from '../models/plugins';
import { RepoPath } from '../models/RepoPath';
import parseLinkText, { determineLinkFile } from './parseLinkText';
import parserInstance from './parserInstance';

export default function buildAst(plugins: Plugins, content: string): Root {
  const parser = parserInstance(plugins);

  try {
    return parser.runSync(parser.parse(content)) as Root;
  } catch (e) {
    throw e;
  }
}

function targetData(
  data: { props: { internalLink?: any } } | null,
): string | null {
  if (data == null) {
    return null;
  }

  return ((data.props || {}).internalLink || {}).target || null;
}

function visit(node: Parent | Content): string[] {
  return [targetData((node.data as any) || null)]
    .concat(
      ((node.children as any) || []).flatMap((n: any): string[] => visit(n)),
    )
    .filter((v): v is string => typeof v === 'string');
}

export function determineDependency(
  repoPath: RepoPath,
  rootNode: Root,
  files: FileBuffer[],
): FileBuffer[] {
  return visit(rootNode)
    .map((item) => parseLinkText(repoPath, item))
    .map((item) => determineLinkFile(item, files))
    .map(
      ({ internalPath: a }) =>
        files.find(({ internalPath: b }) => compareInternalPath(a, b)) ||
        newDoesNotExistFile(a),
    )
    .flatMap((f) => (f ? [f] : []));
}
