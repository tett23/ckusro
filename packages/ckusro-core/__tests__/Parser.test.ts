import buildAst from '../src/Parser/buildAst';
import { buildCkusroConfig } from './__fixtures__';

describe('Parser', () => {
  it('', () => {
    const config = buildCkusroConfig();
    const actual = buildAst(config.plugins, 'test');

    expect(actual).toMatchSnapshot();
  });
});
