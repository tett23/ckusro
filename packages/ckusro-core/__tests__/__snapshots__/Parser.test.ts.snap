// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser  1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 5,
              "line": 1,
              "offset": 4,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "text",
          "value": "test",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 5,
          "line": 1,
          "offset": 4,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 5,
      "line": 1,
      "offset": 4,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
